week2/LogAnalysis.hs

```diff
@@ -0,0 +1,56 @@
+{-# OPTIONS_GHC -Wall #-}
+
+module LogAnalysis where
+
+import Log
+
+-- EXERCISE 1 --
+
+parseMessage :: String -> LogMessage
+parseMessage = 
+  convert . words
+  where convert :: [String] -> LogMessage
+        convert ("I":t:m) = LogMessage Info (read t) (unwords m)
+        convert ("W":t:m) = LogMessage Warning (read t) (unwords m)
+        convert ("E":c:t:m) = LogMessage (Error (read c)) (read t) (unwords m)
+        convert x = Unknown (unwords x)
+
+parse :: String -> [LogMessage]
+parse = map parseMessage . lines
+
+-- EXERCISE 2 --
+
+insert :: LogMessage -> MessageTree -> MessageTree
+insert (Unknown _) tree = tree
+insert newLog Leaf = Node Leaf newLog Leaf
+insert newLog (Node left root right) = 
```

Comment:
You can do pattern-matching here directly if you want to:
```
insert newLog@(LogMessage _ t1 _ ) (Node left root@(LogMessage _ t2 _ ) right) = 
```
you wouldn't cover the case in which `root` is of type `Unknown` - but if your logic for building a `MessageTree` does not allow for that, then you can be relatively safe